staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -156,11 +172,57 @@ type crdInfo struct {
 	storageVersion string
 
 	waitGroup *utilwaitgroup.SafeWaitGroup
+
+	// storageVersionUpdate holds information about the storage version
+	// update issued when the crdInfo was created. The API server uses the
+	// information to decide whether a CR write request should be allowed,
+	// rejected, or blocked.
+	storageVersionUpdate *storageVersionUpdate
 }
 
 // crdStorageMap goes from customresourcedefinition to its storage
 type crdStorageMap map[types.UID]*crdInfo
 
+// storageVersionUpdate holds information about a storage version update,
+// indicating whether the update gets processed, or timed-out.
+type storageVersionUpdate struct {
+	// processedCh is closed by the storage version manager after the
+	// storage version update gets processed (either succeeded or failed).
+	// The API server will unblock and allow CR write requests if this
+	// channel is closed.
+	processedCh <-chan struct{}
+
+	// timeout is the time when the API server will unblock and allow CR
+	// write requests even if the storage version update hasn't been
+	// processed.
+	timeout time.Time
+}
+
+func (i *crdInfo) waitForStorageVersionUpdate(ctx context.Context) error {
+	// StorageVersionAPI feature gate is disabled
+	if i.storageVersionUpdate == nil {
+		return nil
+	}
+
+	// NOTE: currently the graceful CRD deletion waits 1s for in-flight requests
+	// to register themselves to the wait group. Ideally the storage version update should
+	// not cause the requests to miss the 1s window; otherwise the requests may
+	// fail ungracefully (e.g. it may happen if the CRD was deleted immediately after the
+	// first CR request establishes the underlying storage).
+	select {
+	case <-i.storageVersionUpdate.processedCh:
+		return nil
+	case <-ctx.Done():
+		return fmt.Errorf("aborted waiting for CRD storage version update: %v", ctx.Err())
+	// Unblock the requests if the storage version update takes a long time, otherwise
+	// CR requests may stack up and overwhelm the API server.
+	// TODO(roycaihw): benchmark the storage version update latency to adjust the timeout.
+	case <-time.After(time.Until(i.storageVersionUpdate.timeout)):
```

Comment:
if I understand the code as currently written, this will also have us waiting for non-mutating requests (including watches) to terminate, so it can take up to 10 minutes pretty normally.  

It's likely worth
1. an integration test establishing a watch and then attempting some mutating requests to see if I've read properly
2. you may need a different kind of exempt determination to prevent writes, but not interrupt reads.

Reply:
Umm why? we are not waiting for SV updates in [read calls](https://github.com/kubernetes/kubernetes/pull/120582/files#diff-01e7a2059fe62e95ec32e2589c01b7218100de327647c641c334d38f2c5aa06eL365-L372) ?

Reply:
Nevertheless, I added another integration test, can you please check if its testing the scenario you had in mind?

Reply:
> Umm why? we are not waiting for SV updates in [read calls](https://github.com/kubernetes/kubernetes/pull/120582/files#diff-01e7a2059fe62e95ec32e2589c01b7218100de327647c641c334d38f2c5aa06eL365-L372) ?

What we're trying to control is the serialization used to store data into etcd.  Read calls do not store data into etcd, so they can proceed while the write calls need to block until we've updated.

Reply:
Oh, followed the chain down below.  I have no memory of prior state, so I can't comment on 4-month-ago-david.